frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -0,0 +1,142 @@
+package bloop.bsp
+
+import bloop.cli.Commands
+import bloop.engine.{Interpreter, State}
+import bloop.io.AbsolutePath
+import ch.epfl.`scala`.bsp.schema._
+import monix.eval.{Task => MonixTask}
+import ch.epfl.scala.bsp.endpoints
+import com.typesafe.scalalogging.Logger
+import org.langmeta.jsonrpc.{
+  JsonRpcClient,
+  Response => JsonRpcResponse,
+  Services => JsonRpcServices
+}
+import org.langmeta.lsp.Window
+
+class BloopBspServices(callSiteState: State, client: JsonRpcClient, bspLogger: Logger) {
+  final val services = JsonRpcServices.empty
+    .requestAsync(endpoints.Build.initialize)(initialize(_))
+    .notification(endpoints.Build.initialized)(initialized(_))
+    .request(endpoints.Workspace.buildTargets)(buildTargets(_))
+    .requestAsync(endpoints.BuildTarget.compile)(compile(_))
+
+  /**
+   * Creates a logger that will forward all the messages to the underlying bsp client.
+   * It does so via the replication of the `window/showMessage` LSP functionality.
+   */
+  private final val bspForwarderLogger = new bloop.logging.AbstractLogger() {
+    override def name: String = "bsp-logger"
+    override def verbose[T](op: => T): T = op
+    override def ansiCodesSupported(): Boolean = true
+
+    override def debug(msg: String): Unit = {
+      callSiteState.logger.debug(msg)
+      //Window.showMessage.info(msg)(client)
+    }
+
+    override def error(msg: String): Unit = {
+      callSiteState.logger.error(msg)
+      Window.showMessage.error(msg)(client)
+    }
+
+    override def warn(msg: String): Unit = {
+      callSiteState.logger.warn(msg)
+      Window.showMessage.warn(msg)(client)
+    }
+
+    override def trace(t: Throwable): Unit = {
+      callSiteState.logger.trace(t)
+      Window.showMessage.info(t.toString)(client)
```

Comment:
Not sure that traces should be forwarded to the clients, it's an even lower log level than debug.

Reply:
They shouldn't, this was for debugging :smile:.